File: firmware/cam-k230/detector.py
# detector — KPU YOLO person detection on the CanMV-K230, with a motion fallback.
#
# The K230's KPU runs a quantized YOLO .kmodel. CanMV exposes it via the `nncase`
# runtime / the `aidemo`-style `nn` API. If the model file is missing (or the API
# differs on your CanMV build) the detector degrades to a frame-difference motion
# gate so main.py still runs end-to-end. Real wiring is marked TODO(model).

try:
    import nncase_runtime as nn   # CanMV KPU runtime (name varies by image)
    _HAVE_KPU = True
except Exception:
    _HAVE_KPU = False

PERSON_CLASS = 0   # COCO "person"

class PersonDetector:
    def __init__(self, model_path):
        self.model_path = model_path
        self.ready = False
        self._prev = None
        if _HAVE_KPU:
            try:
                # TODO(model): load + build the kmodel KPU runtime here, e.g.
                #   self.kpu = nn.from_oj(open(model_path,'rb').read())
                #   self.kpu.set_input_tensor(...) etc.
                # Then detect() pre-processes the frame to the model input size,
                # runs the KPU, and parses YOLO boxes -> top person score.
                open(model_path, "rb").close()   # presence check only for now
                self.ready = False               # flip True once parsing is wired
            except OSError:
                self.ready = False

    def name(self):
        return "k230-kpu-yolo" if self.ready else "k230-motion-fallback"

    def detect(self, frame):
        """frame: an image object (image.Image). Returns (is_person, confidence)."""
        if self.ready:
            # TODO(model): run KPU + parse boxes; return top person box.
            return (False, 0.0)
        return self._motion(frame)

    def _motion(self, frame):
        # Coarse luma diff: sample a small grid from the frame buffer.
        try:
            buf = frame.bytearray()
        except Exception:
            return (False, 0.0)
        n = 32 * 24
        if len(buf) < n:
            return (False, 0.0)
        stride = len(buf) // n
        cur = bytes(buf[i * stride] for i in range(n))
        if self._prev is None:
            self._prev = cur
            return (False, 0.0)
        changed = sum(1 for a, b in zip(cur, self._prev) if abs(a - b) > 24)
        self._prev = cur
        frac = changed / float(n)
        # Reported as confidence; main maps a positive to kind="motion".
        return (False, frac if frac >= 0.04 else 0.0)

File: firmware/cam-pi/src/detector.cpp
#include "hearth_pi/detector.h"
#include <cstdlib>
#include <cstring>
#include <vector>

namespace hearth {

struct PersonDetector::Impl {
    bool ready = false;
    // motion-stub state: previous coarse luma map
    std::vector<uint8_t> prev;
};

PersonDetector::PersonDetector(std::string modelPath)
    : modelPath_(std::move(modelPath)) { impl_ = new Impl(); }

PersonDetector::~PersonDetector() { delete impl_; }

bool PersonDetector::begin() {
#ifdef HEARTH_HAVE_ONNX
    // TODO(model): create an Ort::Session over modelPath_ (yolov8n), set up the
    // input/output names, and mark ready. detect() then letterboxes the RGB
    // frame to 640x640, runs the session, and parses class-0 (person) boxes.
    // This mirrors the hub's vision pipeline (reuse src/vision if linked).
    impl_->ready = false;   // flip true once the session is wired
#else
    impl_->ready = false;
#endif
    return true;   // never fail startup; degrade to motion
}

const char* PersonDetector::backend() const {
#ifdef HEARTH_HAVE_HAILO
    return impl_->ready ? "hailo" : "motion-stub";
#elif defined(HEARTH_HAVE_ONNX)
    return impl_->ready ? "onnxruntime" : "motion-stub";
#else
    return "motion-stub";
#endif
}

Detection PersonDetector::detect(const Frame& f) {
    Detection d;
    if (impl_->ready) {
        // TODO(model): run inference and return the top person box.
        return d;
    }
    // Motion-stub fallback: 32x24 coarse luma diff over the RGB frame.
    if (!f.rgb || f.width <= 0 || f.height <= 0) return d;
    const int GW = 32, GH = 24, N = GW * GH;
    std::vector<uint8_t> cur(N);
    for (int gy = 0; gy < GH; ++gy)
        for (int gx = 0; gx < GW; ++gx) {
            int px = (gx * f.width) / GW, py = (gy * f.height) / GH;
            const uint8_t* p = f.rgb + (py * f.width + px) * 3;
            cur[gy * GW + gx] = (uint8_t)((p[0] * 30 + p[1] * 59 + p[2] * 11) / 100); // luma
        }
    if (impl_->prev.size() != (size_t)N) { impl_->prev = cur; return d; }
    int changed = 0;
    for (int i = 0; i < N; ++i)
        if (std::abs((int)cur[i] - (int)impl_->prev[i]) > 24) changed++;
    impl_->prev = cur;
    float frac = (float)changed / N;
    // Reported as confidence; the service maps a positive to kind="motion".
    d.person = false;
    d.confidence = (frac >= 0.04f) ? frac : 0.0f;
    return d;
}

} // namespace hearth
